airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py

```diff
@@ -1142,8 +1155,45 @@ def compare_records(
         extra_records: bool,
         ignored_fields: List[str],
         detailed_logger: Logger,
+        configured_catalog: ConfiguredAirbyteCatalog,
     ):
         """Compare records using combination of restrictions"""
+        configured_streams = [stream for stream in configured_catalog.streams if stream.stream.name == stream_name]
+        if len(configured_streams) != 1:
+            raise ValueError(f"Expected exactly one stream matching name {stream_name} but got {len(configured_streams)}")
+
+        configured_stream = configured_streams[0]
+        if configured_stream.stream.source_defined_primary_key:
+            # as part of the migration for relaxing CATs, we are starting only with the streams that defines primary keys
+            expected_primary_keys = list(_extract_primary_keys(expected, configured_stream.stream.source_defined_primary_key))
+            actual_primary_keys = list(_extract_primary_keys(actual, configured_stream.stream.source_defined_primary_key))
```

Comment:
nit: Can we take advantage of the fact your `_extract_primary_keys` function is returning generators to iterate on expected and iterate on a `zip` of both generator and fail if pk do not match (works for `exact_order` only)

```python
for expected_pk, actual_pk in zip((_extract_primary_keys(expected, ...),  _extract_primary_keys(actual, ...):
  assert expected_pk == actual_pk 
```
This could be better when iterating on fat records / long records and fail fast.

Reply:
The drawback is that with this assert, the dev don't have the full picture when they will want to know where was the issue. Maybe all the IDs are there but just not in order. The current assert allows for a clean diff on the whole output. I prefer to take a little bit more time bit have a clearer picture for the dev. Does that make sense? If not, I can use the generator as you suggested

https://private-user-images.githubusercontent.com/3360483/307145330-02cc2821-fd5b-4f61-b02c-139667e36fea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwNTE0ODQsIm5iZiI6MTcwOTA1MTE4NCwicGF0aCI6Ii8zMzYwNDgzLzMwNzE0NTMzMC0wMmNjMjgyMS1mZDViLTRmNjEtYjAyYy0xMzk2NjdlMzZmZWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjdUMTYyNjI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWViOTI0MWVhNTFmZjUxOTc0YjJmMTEwODc1MWIxMjI2NjE2NzFkZmJjMzUzMzFjMzBiY2MwMmQ3MjFiMjI2NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ZuqJoSddnvfLIkF-qrwy2X_jMkGFRfaps4iJ3EU7C8E